### ISSUE TITLE
git-lfs 3.3.0 cannot handle unix-like paths from MSYS2's git anymore

### ISSUE THREAD

**Issue by msarehn** [git-lfs/git-lfs#5250]
### Issue

We're running MSYS2 in our Windows build environments. When running `git lfs pull` in a repository, this now fails with

```
# GIT_TRACE=1 git lfs pull
10:32:29.440933 git.c:750               trace: exec: git-lfs pull
10:32:29.441742 run-command.c:655       trace: run_command: git-lfs pull
10:32:29.486760 trace git-lfs: exec: git 'version'
10:32:29.560580 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'rev-parse' '--git-dir' '--show-toplevel'
10:32:29.621112 trace git-lfs: exec: uname
10:32:29.677833 trace git-lfs: Error running 'git rev-parse': error converting "/builds/5aqPQdE1/0/gitlab/<path to git repo>" to absolute: The system cannot find the path specified.
Error: error converting "/builds/5aqPQdE1/0/gitlab/<path to git repo>"
to absolute: The system cannot find the path specified.
Not in a Git repository.
```

Downgrading to git-lfs 3.2.0 fixes the issue.

Sifting through the ChangeLog, it appears that #5106 might be at fault. And indeed, running git-lfs like

```
# MSYSTEM=MSYS git lfs pull

```

No issue anymore! In MSys2, even in an MINGW shell we're using the MSYS git, so dropping the check for `"MINGW"` seems wrong. Or should this be patched in MSys2?

### Versions

```
# git --version
git version 2.38.1
# git-lfs --version
git-lfs/3.3.0 (GitHub; windows amd64; go 1.19.3)
# pacman -Qi git mingw-w64-x86_64-git-lfs
Name            : git
Version         : 2.38.1-1
Description     : The fast distributed version control system
Architecture    : x86_64
URL             : https://git-scm.com/
Licenses        : GPL2
Groups          : VCS
Provides        : git-core
Depends On      : curl  libpcre2_8  libexpat  libintl  nano  openssh  openssl  perl-Error  perl>=5.14.0
                  perl-Authen-SASL  perl-libwww  perl-MIME-tools  perl-Net-SMTP-SSL  perl-TermReadKey
Optional Deps   : python: various helper scripts
                  subversion: git svn
Required By     : mingw-w64-x86_64-git-lfs
Optional For    : None
Conflicts With  : None
Replaces        : git-core
Installed Size  : 34.92 MiB
Packager        : CI (msys2/msys2-autobuild/9d582e19/3276290976)
Build Date      : Tue, Oct 18, 2022 22:48:09
Install Date    : Mon, Jan 9, 2023 10:39:17
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature

Name            : mingw-w64-x86_64-git-lfs
Version         : 3.3.0-1
Description     : An open source Git extension for versioning large files (mingw-w64)
Architecture    : any
URL             : https://git-lfs.github.com/
Licenses        : MIT
Groups          : None
Provides        : None
Depends On      : git
Optional Deps   : None
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 14.40 MiB
Packager        : CI (msys2/msys2-autobuild/59e8e1af/3590037304)
Build Date      : Thu, Dec 1, 2022 07:21:09
Install Date    : Mon, Jan 9, 2023 10:49:22
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature
```

**Comment by bk2204** [comment]
Hey,

This problem occurs because Git for Windows has a broken `cygpath` and it's indistinguishable from MSys2.  In order for us to treat MSys2 separately from Git for Windows, it needs to return a `uname` value different from Git for Windows.  If not, then this will probably need to be patched on the MSys2 side.

**Comment by msarehn** [comment]
I'm just wondering how this works *at all* on Git for Windows. But thinking about it, Git for Windows must have native Windows path handling, I guess.
MSys2's git does not.

For the time being, this either needs a different way of detecting MSys2 or MSys2 has to revert the patch for their builds.

**Comment by bk2204** [comment]
Personally, I would expect that MSys2 would report `MSYS` in the output of `uname` instead of `MINGW`.  If it reported `MSYS` or `MSYS2`, then this wouldn't be a problem, but #5106 would only affect it if it reported `MINGW`.  Unfortunately, that makes it difficult to reason about what environment is being used.

If there's a portable way to get the environment (à la `lsb_release`) that works on all of Cygwin, MSys, MSys2, and Git for Windows, then we could switch to that instead.  Note that we can't accept any sort of path relative to the root of the Unix-like environment, since we can't resolve such a path without `cygpath` (which might be broken), so we'd need a binary in `PATH` that we could run.

**Comment by msarehn** [comment]
Reading https://github.com/git-for-windows/git/wiki/Install-inside-MSYS2-proper, it might make more sense to instead inspect the output from `git --version`. According to that wiki page, `git --version` should contain the string `.windows`, which the "normal" git doesn't produce.

**Comment by bk2204** [comment]
Are we sure that's true for all versions of Git on Windows that are not MSys, Cygwin, or Mingw?  I ask because there are people who distribute Git with a variety of pieces of software, such as editors and tools like GitHub Desktop, and I'm not sure that they're all going to contain `.windows`.  Similarly, Git for Windows can also be compiled on Unix (which someone might do because it has a few more features) and as a consequence we could end up with a Git for Windows compiled for one of the Unix environments or even Linux.

I think it would be safer here to have an approach which specifically says that on Windows, such and such an environment needs Unix paths (since that's not the default), which is what we're trying to do with `uname`.  Is there an approach that we could take that indicates that better than the current approach?

**Comment by msarehn** [comment]
Good point. The assumption doesn't hold, unfortunately. There seems to be some discrepency between building git-for-windows from a git checkout or from a release tarball. If built from git, the build system generates the version info from the tag (see https://github.com/git-for-windows/git/blob/main/GIT-VERSION-GEN). This one is always guaranteed to include the `.windows` part.  
If built from source tarballs, there is not git tag with `.windows` in it and the usual git version is used as a fallback, not including any special string.

That's unfortunate. For the meantime, downstream reverts the commit in question, but I don't think that this is a viable long-term solution.

Should we leave this issue open until we can come up with something proper?

**Comment by nhatkhai** [comment]
Please fix this
